File: src/infrastructure/external/WhatsAppService.ts
import crypto from 'crypto';
import axios from 'axios';

export interface WhatsAppMessage {
  id: string;
  from: string;
  timestamp: string;
  type: 'text' | 'image' | 'audio' | 'video' | 'document' | 'location';
  text?: {
    body: string;
  };
  image?: {
    id: string;
    mime_type: string;
    sha256: string;
  };
  audio?: {
    id: string;
    mime_type: string;
  };
  location?: {
    latitude: number;
    longitude: number;
    name?: string;
    address?: string;
  };
}

export interface WhatsAppWebhookPayload {
  object: string;
  entry: Array<{
    id: string;
    changes: Array<{
      value: {
        messaging_product: string;
        metadata: {
          display_phone_number: string;
          phone_number_id: string;
        };
        contacts?: Array<{
          profile: {
            name: string;
          };
          wa_id: string;
        }>;
        messages?: WhatsAppMessage[];
        statuses?: Array<{
          id: string;
          status: string;
          timestamp: string;
          recipient_id: string;
        }>;
      };
      field: string;
    }>;
  }>;
}

export interface SendMessageRequest {
  messaging_product: 'whatsapp';
  to: string;
  type: 'text' | 'template';
  text?: {
    body: string;
  };
  template?: {
    name: string;
    language: {
      code: string;
    };
    components?: any[];
  };
}

export class WhatsAppService {
  private readonly accessToken: string;
  private readonly phoneNumberId: string;
  private readonly webhookSecret: string;
  private readonly apiVersion: string = 'v21.0';
  private readonly baseUrl: string;

  constructor() {
    this.accessToken = process.env.WHATSAPP_TOKEN || '';
    this.phoneNumberId = process.env.WHATSAPP_PHONE_NUMBER_ID || '';
    this.webhookSecret = process.env.WHATSAPP_WEBHOOK_SECRET || '';
    this.baseUrl = `https://graph.facebook.com/${this.apiVersion}`;

    if (!this.accessToken) {
      console.warn('WHATSAPP_TOKEN not configured');
    }
    if (!this.phoneNumberId) {
      console.warn('WHATSAPP_PHONE_NUMBER_ID not configured');
    }
    if (!this.webhookSecret) {
      console.warn('WHATSAPP_WEBHOOK_SECRET not configured');
    }
  }

  /**
   * Valida a assinatura do webhook do WhatsApp
   */
  public validateSignature(payload: string, signature: string): boolean {
    if (!this.webhookSecret) {
      console.warn('Webhook secret not configured, skipping signature validation');
      return true; // Em desenvolvimento, permitir sem validação
    }

    try {
      const expectedSignature = crypto
        .createHmac('sha256', this.webhookSecret)
        .update(payload)
        .digest('hex');

      const receivedSignature = signature.replace('sha256=', '');
      
      return crypto.timingSafeEqual(
        Buffer.from(expectedSignature, 'hex'),
        Buffer.from(receivedSignature, 'hex')
      );
    } catch (error) {
      console.error('Error validating webhook signature:', error);
      return false;
    }
  }

  /**
   * Processa o payload do webhook
   */
  public processWebhookPayload(payload: WhatsAppWebhookPayload): WhatsAppMessage[] {
    const messages: WhatsAppMessage[] = [];

    if (payload.object !== 'whatsapp_business_account') {
      console.warn('Received webhook for unknown object:', payload.object);
      return messages;
    }

    payload.entry?.forEach(entry => {
      entry.changes?.forEach(change => {
        if (change.field === 'messages' && change.value.messages) {
          messages.push(...change.value.messages);
        }
      });
    });

    return messages;
  }

  /**
   * Envia uma mensagem de texto via WhatsApp
   */
  public async sendTextMessage(to: string, message: string): Promise<any> {
    if (!this.accessToken || !this.phoneNumberId) {
      throw new Error('WhatsApp credentials not configured');
    }

    const payload: SendMessageRequest = {
      messaging_product: 'whatsapp',
      to: to.replace(/\D/g, ''), // Remove caracteres não numéricos
      type: 'text',
      text: {
        body: message
      }
    };

    try {
      const response = await axios.post(
        `${this.baseUrl}/${this.phoneNumberId}/messages`,
        payload,
        {
          headers: {
            'Authorization': `Bearer ${this.accessToken}`,
            'Content-Type': 'application/json'
          }
        }
      );

      console.log('Message sent successfully:', response.data);
      return response.data;
    } catch (error: any) {
      console.error('Error sending WhatsApp message:', error.response?.data || error.message);
      throw error;
    }
  }

  /**
   * Envia uma mensagem de template via WhatsApp
   */
  public async sendTemplateMessage(
    to: string, 
    templateName: string, 
    languageCode: string = 'pt_BR',
    components?: any[]
  ): Promise<any> {
    if (!this.accessToken || !this.phoneNumberId) {
      throw new Error('WhatsApp credentials not configured');
    }

    const payload: SendMessageRequest = {
      messaging_product: 'whatsapp',
      to: to.replace(/\D/g, ''),
      type: 'template',
      template: {
        name: templateName,
        language: {
          code: languageCode
        },
        components: components || []
      }
    };

    try {
      const response = await axios.post(
        `${this.baseUrl}/${this.phoneNumberId}/messages`,
        payload,
        {
          headers: {
            'Authorization': `Bearer ${this.accessToken}`,
            'Content-Type': 'application/json'
          }
        }
      );

      console.log('Template message sent successfully:', response.data);
      return response.data;
    } catch (error: any) {
      console.error('Error sending WhatsApp template:', error.response?.data || error.message);
      throw error;
    }
  }

  /**
   * Marca uma mensagem como lida
   */
  public async markAsRead(messageId: string): Promise<any> {
    if (!this.accessToken || !this.phoneNumberId) {
      throw new Error('WhatsApp credentials not configured');
    }

    try {
      const response = await axios.post(
        `${this.baseUrl}/${this.phoneNumberId}/messages`,
        {
          messaging_product: 'whatsapp',
          status: 'read',
          message_id: messageId
        },
        {
          headers: {
            'Authorization': `Bearer ${this.accessToken}`,
            'Content-Type': 'application/json'
          }
        }
      );

      return response.data;
    } catch (error: any) {
      console.error('Error marking message as read:', error.response?.data || error.message);
      throw error;
    }
  }

  /**
   * Obtém informações do perfil de um contato
   */
  public async getContactProfile(phoneNumber: string): Promise<any> {
    if (!this.accessToken) {
      throw new Error('WhatsApp access token not configured');
    }

    try {
      const response = await axios.get(
        `${this.baseUrl}/${phoneNumber.replace(/\D/g, '')}`,
        {
          headers: {
            'Authorization': `Bearer ${this.accessToken}`
          }
        }
      );

      return response.data;
    } catch (error: any) {
      console.error('Error getting contact profile:', error.response?.data || error.message);
      throw error;
    }
  }

  /**
   * Verifica se as credenciais estão configuradas
   */
  public isConfigured(): boolean {
    // Verificar se os tokens são placeholders
    const isPlaceholderToken = this.accessToken.includes('your-') || this.accessToken.includes('token-here');
    const isPlaceholderPhoneId = this.phoneNumberId.includes('your-') || this.phoneNumberId.includes('id-here');
    
    if (isPlaceholderToken || isPlaceholderPhoneId) {
      console.log('🔧 WhatsApp tokens are placeholders, using simulation mode');
      return false; // Usar modo simulação
    }
    
    return !!(this.accessToken && this.phoneNumberId);
  }

  /**
   * Obtém status da configuração
   */
  public getConfigurationStatus(): {
    accessToken: boolean;
    phoneNumberId: boolean;
    webhookSecret: boolean;
  } {
    return {
      accessToken: !!this.accessToken,
      phoneNumberId: !!this.phoneNumberId,
      webhookSecret: !!this.webhookSecret
    };
  }
}

File: python_datacompass/app/infrastructure/external/whatsapp_service.py
"""
WhatsApp Business API service using PyWA.
"""

import hmac
import hashlib
from typing import Any, Dict, List, Optional

from pywa import WhatsApp
from pywa.types import Message, MessageType, MessageStatus
from pywa.errors import WhatsAppError

from app.core.config import settings
from app.core.logging import LoggerMixin


class WhatsAppService(LoggerMixin):
    """WhatsApp Business API service."""
    
    def __init__(self):
        self.client: Optional[WhatsApp] = None
        self._initialize_client()
    
    def _initialize_client(self) -> None:
        """Initialize WhatsApp client."""
        if not settings.whatsapp_token or not settings.whatsapp_phone_number_id:
            self.logger.warning("WhatsApp credentials not configured - using simulation mode")
            return
        
        try:
            self.client = WhatsApp(
                phone_id=settings.whatsapp_phone_number_id,
                token=settings.whatsapp_token,
                verify_token=settings.whatsapp_webhook_verify_token,
            )
            self.logger.info("WhatsApp client initialized successfully")
        except Exception as e:
            self.logger.error(f"Failed to initialize WhatsApp client: {e}")
            self.client = None
    
    def is_configured(self) -> bool:
        """Check if WhatsApp is properly configured."""
        return (
            self.client is not None and
            settings.whatsapp_token and
            settings.whatsapp_phone_number_id and
            not settings.whatsapp_token.startswith("your-") and
            not settings.whatsapp_phone_number_id.startswith("your-")
        )
    
    def validate_signature(self, payload: str, signature: str) -> bool:
        """Validate WhatsApp webhook signature."""
        if not settings.whatsapp_webhook_secret:
            self.logger.warning("Webhook secret not configured, skipping signature validation")
            return True
        
        try:
            expected_signature = hmac.new(
                settings.whatsapp_webhook_secret.encode('utf-8'),
                payload.encode('utf-8'),
                hashlib.sha256
            ).hexdigest()
            
            received_signature = signature.replace('sha256=', '')
            
            return hmac.compare_digest(expected_signature, received_signature)
        except Exception as e:
            self.logger.error(f"Error validating webhook signature: {e}")
            return False
    
    async def send_text_message(self, to: str, message: str) -> Dict[str, Any]:
        """Send a text message via WhatsApp."""
        if not self.is_configured():
            self.logger.info(f"[SIMULATION] Sending message to {to}: {message}")
            return {
                "success": True,
                "message_id": f"sim-{hash(message)}",
                "simulated": True
            }
        
        try:
            # Clean phone number
            clean_number = ''.join(filter(str.isdigit, to))
            
            # Send message using PyWA
            result = await self.client.send_message(
                to=clean_number,
                text=message
            )
            
            self.logger.info(f"Message sent successfully to {to}")
            return {
                "success": True,
                "message_id": result.id,
                "whatsapp_response": result
            }
            
        except WhatsAppError as e:
            self.logger.error(f"WhatsApp API error: {e}")
            return {
                "success": False,
                "error": str(e)
            }
        except Exception as e:
            self.logger.error(f"Unexpected error sending message: {e}")
            return {
                "success": False,
                "error": str(e)
            }
    
    async def send_template_message(
        self, 
        to: str, 
        template_name: str, 
        language_code: str = "pt_BR",
        components: Optional[List[Dict[str, Any]]] = None
    ) -> Dict[str, Any]:
        """Send a template message via WhatsApp."""
        if not self.is_configured():
            self.logger.info(f"[SIMULATION] Sending template {template_name} to {to}")
            return {
                "success": True,
                "message_id": f"sim-template-{hash(template_name)}",
                "simulated": True
            }
        
        try:
            clean_number = ''.join(filter(str.isdigit, to))
            
            # Send template message using PyWA
            result = await self.client.send_template(
                to=clean_number,
                template=template_name,
                language=language_code,
                components=components or []
            )
            
            self.logger.info(f"Template message sent successfully to {to}")
            return {
                "success": True,
                "message_id": result.id,
                "whatsapp_response": result
            }
            
        except WhatsAppError as e:
            self.logger.error(f"WhatsApp template API error: {e}")
            return {
                "success": False,
                "error": str(e)
            }
        except Exception as e:
            self.logger.error(f"Unexpected error sending template: {e}")
            return {
                "success": False,
                "error": str(e)
            }
    
    async def send_document(
        self, 
        to: str, 
        file_path: str, 
        caption: Optional[str] = None
    ) -> Dict[str, Any]:
        """Send a document via WhatsApp."""
        if not self.is_configured():
            self.logger.info(f"[SIMULATION] Sending document {file_path} to {to}")
            return {
                "success": True,
                "message_id": f"sim-doc-{hash(file_path)}",
                "simulated": True
            }
        
        try:
            clean_number = ''.join(filter(str.isdigit, to))
            
            # Determine if it's an image or document
            import os
            file_extension = os.path.splitext(file_path)[1].lower()
            image_extensions = ['.png', '.jpg', '.jpeg', '.gif']
            
            if file_extension in image_extensions:
                # Send as image
                result = await self.client.send_image(
                    to=clean_number,
                    image=file_path,
                    caption=caption
                )
            else:
                # Send as document
                result = await self.client.send_document(
                    to=clean_number,
                    document=file_path,
                    caption=caption
                )
            
            self.logger.info(f"Document sent successfully to {to}")
            return {
                "success": True,
                "message_id": result.id,
                "whatsapp_response": result
            }
            
        except WhatsAppError as e:
            self.logger.error(f"WhatsApp document API error: {e}")
            return {
                "success": False,
                "error": str(e)
            }
        except Exception as e:
            self.logger.error(f"Unexpected error sending document: {e}")
            return {
                "success": False,
                "error": str(e)
            }
    
    async def mark_as_read(self, message_id: str) -> bool:
        """Mark a message as read."""
        if not self.is_configured():
            self.logger.info(f"[SIMULATION] Marking message {message_id} as read")
            return True
        
        try:
            await self.client.mark_as_read(message_id)
            return True
        except Exception as e:
            self.logger.error(f"Error marking message as read: {e}")
            return False
    
    async def get_profile(self, phone_number: str) -> Optional[Dict[str, Any]]:
        """Get WhatsApp profile information."""
        if not self.is_configured():
            self.logger.info(f"[SIMULATION] Getting profile for {phone_number}")
            return {
                "name": "Simulated User",
                "simulated": True
            }
        
        try:
            clean_number = ''.join(filter(str.isdigit, phone_number))
            profile = await self.client.get_profile(clean_number)
            return profile
        except Exception as e:
            self.logger.error(f"Error getting profile: {e}")
            return None
    
    def get_configuration_status(self) -> Dict[str, bool]:
        """Get WhatsApp configuration status."""
        return {
            "access_token": bool(settings.whatsapp_token),
            "phone_number_id": bool(settings.whatsapp_phone_number_id),
            "webhook_secret": bool(settings.whatsapp_webhook_secret),
            "configured": self.is_configured()
        }
    
    def process_webhook_message(self, message: Message) -> Dict[str, Any]:
        """Process a webhook message from WhatsApp."""
        return {
            "id": message.id,
            "from": message.from_user.wa_id,
            "timestamp": message.timestamp,
            "type": message.type.value if message.type else "unknown",
            "text": message.text.body if message.text else None,
            "document": {
                "id": message.document.id,
                "filename": message.document.filename,
                "mime_type": message.document.mime_type
            } if message.document else None,
            "image": {
                "id": message.image.id,
                "mime_type": message.image.mime_type
            } if message.image else None,
            "audio": {
                "id": message.audio.id,
                "mime_type": message.audio.mime_type
            } if message.audio else None
        }